### ISSUE TITLE
`vast-front` verifier failure on elaborated types

### ISSUE THREAD

**Issue by surovic** [trailofbits/vast#286]
Attached is an archive containing `bug.c` (the input file) and output with the verifier disabled via `-vast-disable-vast-verifier`.

input:

```c
struct dentry_operations {
    char *d_dname;
};

struct dentry {
    const struct dentry_operations *d_op;
};

struct path {
    struct dentry *dentry;
};

int d_path(const struct path *path)
{
    if (path->dentry->d_op && path->dentry->d_op->d_dname)
        return 1;
    
    return 0;
}
```

tool invocation:
```sh
vast-front -vast-emit-mlir=hl bug.c -o -
```

error:
```
loc(fused<#meta.id<13>>[unknown]): error: 'hl.ref' op inferred type(s) '!hl.lvalue<!hl.ptr<!hl.record<"path",  const >>>' are incompatible with return type(s) of operation '!hl.lvalue<!hl.ptr<!hl.elaborated<!hl.record<"path">,  const >>>'
[vast] codegen: module verification error before running vast passes
```

[bug.zip](https://github.com/trailofbits/vast/files/11263197/bug.zip)

**Comment by surovic** [comment]
So I think I've located the cause of the bug, but I have no idea what to do with it. The following is the code on which the verifier fails.

```MLIR
"hl.func"() ({
  ^bb0(%arg0: !hl.ptr<!hl.record<"path",  const >> loc(fused<#meta.id<10>>[unknown])):
    "hl.if"() ({
      %1 = "hl.bin.land"() ({
        %2 = "hl.ref"(%arg0) : (!hl.ptr<!hl.record<"path",  const >>) -> !hl.lvalue<!hl.ptr<!hl.elaborated<!hl.record<"path">,  const >>> loc(#loc11)
```

`hl.ref` should have the same result as its operand. The operand to `hl.ref` is `%arg0` whose type is derived from the function prototype in [TypeInfo.cpp](https://github.com/trailofbits/vast/blob/6843aafdaba6dd4555f34e6295b8044437c3cb71/lib/vast/CodeGen/TypeInfo.cpp#L46).

```c++
auto fty = fn->getType()->getCanonicalTypeUnqualified();
```

Output of `fn->getType()->dump()`:

```
FunctionProtoType 0x5614d3591d80 'int (const struct path *)' cdecl
|-BuiltinType 0x5614d340c360 'int'
`-PointerType 0x5614d3591c90 'const struct path *'
  `-QualType 0x5614d3591c31 'const struct path' const
    `-ElaboratedType 0x5614d3591c30 'struct path' sugar
      `-RecordType 0x5614d3591ab0 'struct path'
        `-Record 0x5614d3591a28 'path'
```

Output of `fty.dump()`:

```
FunctionProtoType 0x5614d3591d50 'int (const struct path *)' cdecl
|-BuiltinType 0x5614d340c360 'int'
`-PointerType 0x5614d3591c60 'const struct path *'
  `-QualType 0x5614d3591ab1 'const struct path' const
    `-RecordType 0x5614d3591ab0 'struct path'
      `-Record 0x5614d3591a28 'path'
```

From the surrounding code in [TypeInfo.cpp](https://github.com/trailofbits/vast/blob/6843aafdaba6dd4555f34e6295b8044437c3cb71/lib/vast/CodeGen/TypeInfo.cpp) it seems like there's a strong requirement on the type to be canonical, which I think implies that it cannot contain an `ElaboratedType`.

What do? :thinking:

**Comment by Jezurko** [comment]
I started working on a fix in [elab-type](https://github.com/trailofbits/vast/tree/elab-type) branch.
We probably don't want to drop the elaborate type info.

I tried to make a dirty patch (e.g. just changing the `using` directives in `include/vast/Translation/Types.hpp`) to stop that behaviour - mainly getting rid of the canonical type and using `QualType/FunctionType` instead.
But it seems there is also another place where we are dropping the information? Not sure if I only missed it at some random place or if there is a completely different place where it also happens. (Or maybe I'm approaching it completely wrong? Can't dismiss that option)
